File: autotest/OldTests/src/test/java/com/tle/webtests/test/myresources/favourites/FavouriteItemsTest.java
package com.tle.webtests.test.myresources.favourites;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertFalse;
import static org.testng.Assert.assertTrue;

import com.tle.webtests.framework.TestInstitution;
import com.tle.webtests.pageobject.portal.MenuSection;
import com.tle.webtests.pageobject.searching.FavouriteItemDialog;
import com.tle.webtests.pageobject.searching.FavouriteItemsPage;
import com.tle.webtests.pageobject.searching.FavouritesPage;
import com.tle.webtests.pageobject.searching.ItemListPage;
import com.tle.webtests.pageobject.searching.ItemSearchResult;
import com.tle.webtests.pageobject.searching.SearchPage;
import com.tle.webtests.pageobject.viewitem.SummaryPage;
import com.tle.webtests.pageobject.wizard.ContributePage;
import com.tle.webtests.pageobject.wizard.WizardPageTab;
import com.tle.webtests.test.AbstractCleanupTest;
import java.util.Optional;
import org.testng.annotations.Test;
import testng.annotation.OldUIOnly;

@TestInstitution("vanilla")
@OldUIOnly
public class FavouriteItemsTest extends AbstractCleanupTest {

  @Test
  public void fromItemSummary() {
    logon("AutoTest", "automated");
    String itemName = context.getFullName("testfromsummary");

    // Add Item
    createItem(itemName);

    // Add to favourites from summary page.
    addFavouriteFromSummary(itemName);

    // Check that it appears on the Favourite Items page
    assertItemVisibleOnFavourites(itemName);

    // Remove it from favourites from summary page.
    removeFavouriteFromSummary(itemName);
  }

  @Test
  public void fromSearchResults() {
    logon("AutoTest", "automated");
    String itemName = context.getFullName("testfromsearchresult");

    // Add Item
    createItem(itemName);

    // Add it to favourites from search results.
    addFavouriteFromSearchPage(itemName);

    // Check that it appears on the Favourite Items page
    assertItemVisibleOnFavourites(itemName);

    // Remove Favourite with link on search result
    removeFavouriteFromSearchPage(itemName);
  }

  @Test
  public void testFavouriteTagSearch() {
    logon("AutoTest", "automated");
    String name1 = "testtagsearch1";
    String itemName1 = context.getFullName(name1);
    String name2 = "testtagsearch2";
    String itemName2 = context.getFullName(name2);
    String tag1 = "tagnumberone";
    String tag2 = "tagnumbertwo,three;four five";

    // Add Item 1 and Item 2
    createItem(itemName1);
    createItem(itemName2);

    // Search for item1 and add it to favourites with tag1
    addFavouriteFromSearchPage(itemName1, tag1);

    // Search for item2 and add it to favourites with tag2
    addFavouriteFromSearchPage(itemName2, tag2);

    // Goto favourites page and assert both items are present before filtering
    FavouriteItemsPage favs = new FavouritesPage(context).load().items();
    assertTrue(favs.results().doesResultExist(itemName1));
    assertTrue(favs.results().doesResultExist(itemName2));

    // Search favs with tag/keyword filter and verify which items are returned
    ItemListPage favresults = favs.search(tag1);
    assertTrue(favresults.doesResultExist(itemName1, 1));
    assertFalse(favresults.doesResultExist(itemName2));

    favresults = favs.search("tagnumbertwo");
    assertTrue(favresults.doesResultExist(itemName2, 1));
    assertFalse(favresults.doesResultExist(itemName1));

    favresults = favs.search("three");
    assertTrue(favresults.doesResultExist(itemName2, 1));
    assertFalse(favresults.doesResultExist(itemName1));

    favresults = favs.search("four");
    assertTrue(favresults.doesResultExist(itemName2, 1));
    assertFalse(favresults.doesResultExist(itemName1));

    favresults = favs.search("five");
    assertTrue(favresults.doesResultExist(itemName2, 1));
    assertFalse(favresults.doesResultExist(itemName1));

    favresults = favs.search("three four five");
    assertTrue(favresults.doesResultExist(itemName2, 1));
    assertFalse(favresults.doesResultExist(itemName1));
  }

  @Test
  public void testVersionFavourites() {
    logon("AutoTest", "automated");
    String itemName = context.getFullName("version");
    String thisVersionTag = "thisversion";
    String latestVersionTag = "latestversion";

    createItem(itemName);

    // Favourite v1 as "This version"
    addFavouriteWithVersion(itemName, thisVersionTag, false);

    // Create v2
    createItemNewVersion(itemName);

    // Favourite v2 as "Latest version"
    addFavouriteWithVersion(itemName, latestVersionTag, true);

    // Verify 'This Version' points to v1 and 'Latest Version' points to v2
    assertFavouriteVersion(thisVersionTag, itemName, 1);
    assertFavouriteVersion(latestVersionTag, itemName, 2);

    // Create v3
    createItemNewVersion(itemName);

    // Verify "Latest version" now points to v3
    assertFavouriteVersion(latestVersionTag, itemName, 3);
  }

  @Test
  public void testNoResults() {
    logon("TLE_ADMINISTRATOR", testConfig.getAdminPassword());

    assertNoFavouriteResults();
  }

  @Test
  public void testAutoLoggedIn() {
    logout().autoLogin(); // Ensure logged out

    // Assert there is no favourite menu
    MenuSection menu = new MenuSection(context).get();
    assertFalse(menu.hasMenuOption("Favourites"));

    // Assert there is no add search to favs button
    SearchPage searchPage = menu.clickMenu("Search", new SearchPage(context));
    assertFalse(searchPage.hasFavouriteSearchOption());

    // Assert there is no add to favs button on item summary
    searchPage.results().getResult(1).clickTitle();
    SummaryPage summaryPage = new SummaryPage(context);
    assertFalse(summaryPage.hasFavouriteOption());
  }

  /** Creates and publishes a basic test item with the provided name. */
  private void createItem(String itemName) {
    WizardPageTab wizard =
        new ContributePage(context).load().openWizard(GENERIC_TESTING_COLLECTION);
    wizard.editbox(1, itemName);
    wizard.save().publish();
  }

  /** Adds an item to favourites from the summary page and asserts success. */
  private void addFavouriteFromSummary(String itemName) {
    SummaryPage itemSummary =
        new SearchPage(context)
            .load()
            .exactQuery(itemName)
            .getResultForTitle(itemName, 1)
            .viewSummary();
    itemSummary.addToFavourites().clickAdd();
    assertTrue(itemSummary.isFavouriteItem());
  }

  /** Adds an item to favourites from the search page and asserts success. */
  private void addFavouriteFromSearchPage(String itemName, String... tags) {
    SearchPage searchPage = new SearchPage(context).load();
    ItemListPage results = searchPage.exactQuery(itemName);
    ItemSearchResult resultForTitle = results.getResultForTitle(itemName, 1);
    FavouriteItemDialog<ItemSearchResult> fav = resultForTitle.addToFavourites();
    Optional.ofNullable(tags).filter(t -> t.length > 0).map(t -> t[0]).ifPresent(fav::setTags);
    fav.clickAdd();

    assertTrue(resultForTitle.isFavouriteItem());
  }

  /** Removes an item from favourites from the summary page and asserts success. */
  private void removeFavouriteFromSummary(String itemName) {
    SummaryPage itemSummary =
        new SearchPage(context)
            .load()
            .exactQuery(itemName)
            .getResultForTitle(itemName, 1)
            .viewSummary();
    itemSummary.removeFavourite();
    itemSummary.get();
    assertFalse(itemSummary.isFavouriteItem());
  }

  /** Removes an item from favourites from the search page and asserts success. */
  private void removeFavouriteFromSearchPage(String itemName) {
    SearchPage searchPage = new SearchPage(context).load();
    ItemListPage results = searchPage.exactQuery(itemName);
    ItemSearchResult resultForTitle = results.getResultForTitle(itemName, 1);
    resultForTitle.removeFavourite();

    results = new SearchPage(context).load().exactQuery(itemName);
    resultForTitle = results.getResultForTitle(itemName, 1);
    assertFalse(resultForTitle.isFavouriteItem());
  }

  /** Asserts an item is visible on the favourites page, optionally filtering by tag. */
  private void assertItemVisibleOnFavourites(String itemName, String... tags) {
    FavouriteItemsPage favs = new FavouritesPage(context).load().items();
    boolean hasItem =
        Optional.ofNullable(tags)
            .filter(t -> t.length > 0)
            .map(t -> favs.search(t[0]).doesResultExist(itemName, 1))
            .orElseGet(() -> favs.results().doesResultExist(itemName));

    assertTrue(hasItem);
  }

  /** Asserts that the favourites page shows no results. */
  private void assertNoFavouriteResults() {
    FavouriteItemsPage favs = new FavouritesPage(context).load().items();
    assertFalse(favs.hasResults());
  }

  /** Adds a version-specific favourite from the search page. */
  private void addFavouriteWithVersion(String itemName, String tag, boolean latestVersion) {
    SearchPage searchPage = new SearchPage(context).load();
    ItemListPage results = searchPage.exactQuery(itemName);
    assertTrue(results.doesResultExist(itemName, 1));

    ItemSearchResult resultForTitle = results.getResultForTitle(itemName, 1);
    resultForTitle.addToFavourites().setTags(tag).setLatestVersion(latestVersion).clickAdd();
  }

  /** Creates a new version of an item from its summary page. */
  private void createItemNewVersion(String itemName) {
    WizardPageTab wizard = SearchPage.searchAndView(context, itemName).adminTab().newVersion();
    wizard.save().publish();
  }

  /** Asserts a favourite points to the correct item version. */
  private void assertFavouriteVersion(String tag, String itemName, int expectedVersion) {
    FavouriteItemsPage favs = new FavouritesPage(context).load().items();
    ItemListPage favresults = favs.search(tag);
    SummaryPage summary = favresults.getResult(1).viewSummary();
    assertEquals(expectedVersion, summary.getItemId().getVersion());
  }
}

File: autotest/OldTests/src/test/scala/io/github/openequella/favourites/FavouriteItemsTest.scala
package io.github.openequella.favourites

import com.tle.webtests.framework.TestInstitution
import com.tle.webtests.pageobject.portal.MenuSection
import com.tle.webtests.pageobject.wizard.ContributePage
import com.tle.webtests.test.AbstractCleanupAutoTest
import io.github.openequella.pages.favourites.FavouritesPage
import io.github.openequella.pages.search.NewSearchPage
import org.testng.Assert.{assertEquals, assertFalse, assertTrue}
import org.testng.annotations.Test
import testng.annotation.NewUIOnly

@TestInstitution("fiveo")
@NewUIOnly
class FavouriteItemsTest extends AbstractCleanupAutoTest {

  @Test(description = "Verify adding and removing an item from favourites on the item summary page")
  def fromItemSummary(): Unit = {
    val itemName = context.getFullName("summary")

    createTestItem(itemName)
    addFavouriteFromSummaryPage(itemName)
    assertItemInFavouritesPage(itemName)
    removeFavouriteFromSummaryPage(itemName)
  }

  @Test(description =
    "Verify adding and removing an item from favourites on the search results page"
  )
  def fromSearchResults(): Unit = {
    val itemName = context.getFullName("search")

    createTestItem(itemName)
    addFavouriteFromSearchPage(itemName)
    assertItemInFavouritesPage(itemName)
    removeFavouriteFromSearchPage(itemName)
  }

  @Test(description = "Verify favourites options are not shown when not logged in / auto logged in")
  def testAutoLoggedIn(): Unit = {
    logout().autoLogin()

    // Verify `Favourites` menu option is not shown
    val menu = new MenuSection(context).get
    assertFalse(
      menu.hasMenuOption("Favourites"),
      "Favourites menu should not be present for auto-logged in users"
    )

    // Open Item summary page for the first item and verify that add/remove favourite button is not shown
    val searchPage = new NewSearchPage(context).load()
    searchPage.waitForInitialSearchResult()
    val firstItem   = searchPage.getItemNameByIndex(0)
    val summaryPage = searchPage.selectItem(firstItem)
    assertFalse(
      summaryPage.hasFavouriteOption,
      "Favourite option should not be available on summary page"
    )
  }

  @Test(description = "Verify version specific favourites (This Version vs Latest Version)")
  def testVersionFavourites(): Unit = {
    val itemName         = context.getFullName("version")
    val tagThisVersion   = "thisversion"
    val tagLatestVersion = "latestversion"

    // Create Item (Version 1)
    createTestItem(itemName)

    // Favourite v1 as "This version"
    val searchPage = new NewSearchPage(context).load()

    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)
    searchPage.addItemToFavourites(itemName, Array(tagThisVersion), false)

    searchPage.newSearch();

    // Create Version 2
    createItemNewVersion(itemName)

    searchPage.load()

    // Favourite v2 as "Latest version"
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)
    searchPage.addItemToFavourites(itemName, Array(tagLatestVersion), true)

    // Verify 'This Version' points to v1 and 'Latest Version' points to v2
    assertFavouriteVersion(tagThisVersion, itemName, 1)
    assertFavouriteVersion(tagLatestVersion, itemName, 2)

    // Create Version 3
    createItemNewVersion(itemName)

    // Verify 'Latest Version' now points to v3
    assertFavouriteVersion(tagLatestVersion, itemName, 3)
  }

  /** Loads the Favourites page, optionally configuring it to show all versions.
    *
    * @param allVersions
    *   when true, expands the refine panel and selects 'All' status values.
    */
  private def loadFavouritesPage(allVersions: Boolean = false): FavouritesPage = {
    val favouritesPage = new FavouritesPage(context).load()
    if (allVersions) {
      favouritesPage.expandRefineControlPanel()
      favouritesPage.selectStatus(true)
      favouritesPage.waitForSearchCompleted()
    }
    favouritesPage
  }

  /** Creates and publishes a basic test item with the provided name.
    *
    * @param itemName
    *   name to use for the contributed item.
    */
  private def createTestItem(itemName: String): Unit = {
    val wizard = new ContributePage(context).load().openWizard("Basic Items")
    wizard.editbox(1, itemName)
    wizard.save().publish()
  }

  /** Navigates to the item summary and clicks the add-to-favourites button.
    *
    * @param itemName
    *   name of the item to locate in search before opening summary.
    */
  private def addFavouriteFromSummaryPage(itemName: String): Unit = {
    val searchPage = new NewSearchPage(context).load()
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)

    val summaryPage = searchPage.selectItem(itemName)
    summaryPage.addToFavourites().clickAdd()
  }

  /** Navigates to the item summary and removes the item from favourites, asserting the summary no
    * longer reports the item as favourited.
    *
    * @param itemName
    *   name of the item whose favourite flag should be removed.
    */
  private def removeFavouriteFromSummaryPage(itemName: String): Unit = {
    val searchPage = new NewSearchPage(context).load()
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)

    val summaryPage = searchPage.selectItem(itemName)
    summaryPage.removeFavourite()
    assertFalse(summaryPage.isFavouriteItem)
  }

  /** Adds the given item to favourites from the search results list.
    *
    * @param itemName
    *   name of the item to favourite in the current search results.
    */
  private def addFavouriteFromSearchPage(itemName: String): Unit = {
    val searchPage = new NewSearchPage(context).load()
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)
    searchPage.addItemToFavourites(itemName)
  }

  /** Removes the given item from favourites in the search results and asserts it is no longer
    * marked as a favourite in the list.
    *
    * @param itemName
    *   name of the item to unfavourite from the search results.
    */
  private def removeFavouriteFromSearchPage(itemName: String): Unit = {
    val searchPage = new NewSearchPage(context).load()
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)
    searchPage.removeFavouriteFromSearchResult(itemName)
    assertFalse(searchPage.isItemFavourited(itemName))
  }

  /** Loads the Favourites page and asserts that the specified item appears in the results.
    *
    * @param itemName
    *   name of the item expected to be present on the Favourites page.
    */
  private def assertItemInFavouritesPage(itemName: String): Unit = {
    val favouritesPage = loadFavouritesPage()
    favouritesPage.waitForSearchCompleted()
    assertTrue(favouritesPage.hasItem(itemName))
  }

  /** Creates a new version of the specified item via the summary page.
    *
    * @param itemName
    *   name of the item for which a new version should be created.
    */
  private def createItemNewVersion(itemName: String): Unit = {
    val searchPage = new NewSearchPage(context).load()
    searchPage.changeQuery(itemName)
    searchPage.waitForSearchCompleted(1)

    val summaryPage = searchPage.selectItem(itemName)
    val wizard      = summaryPage.adminTab().newVersion()
    wizard.save().publish()
  }

  /** Verifies that a favourite tagged with the given name resolves to the expected item version on
    * the Favourites page.
    *
    * @param tagName
    *   tag applied when the favourite entry was created.
    * @param itemName
    *   name of the favourited item.
    * @param expectedVersion
    *   item version expected when opening the favourite.
    */
  private def assertFavouriteVersion(
      tagName: String,
      itemName: String,
      expectedVersion: Int
  ): Unit = {
    val favouritesPage = loadFavouritesPage(allVersions = true)

    favouritesPage.changeQuery(tagName)
    favouritesPage.waitForSearchCompleted(1)
    assertTrue(favouritesPage.hasItem(itemName))

    val summaryPage = favouritesPage.selectItem(itemName)
    assertEquals(summaryPage.getItemId.getVersion, expectedVersion)
  }
}
